File: python/src/models/transformer.py
"""Proper transformer circuit model with primary and secondary windings.

This model represents a real transformer with:
- Primary winding (input side)
- Secondary winding (output side)
- Turns ratio for voltage transformation
- Magnetizing inductance
- Leakage inductances
- Winding resistances
"""

import numpy as np


class Transformer:
    """Real transformer model with primary/secondary windings and voltage transformation."""
    
    def __init__(self, voltage_primary, frequency, turns_ratio, 
                 inductance_mag, resistance_primary, resistance_secondary):
        """Initialize the transformer.
        
        Args:
            voltage_primary: Primary RMS voltage in volts
            frequency: Frequency in Hz
            turns_ratio: N1/N2 (primary turns / secondary turns)
            inductance_mag: Magnetizing inductance in henries
            resistance_primary: Primary winding resistance in ohms
            resistance_secondary: Secondary winding resistance in ohms
        """
        if voltage_primary <= 0:
            raise ValueError(f"Primary voltage must be positive, got {voltage_primary}")
        if frequency <= 0:
            raise ValueError(f"Frequency must be positive, got {frequency}")
        if turns_ratio <= 0:
            raise ValueError(f"Turns ratio must be positive, got {turns_ratio}")
        if inductance_mag < 0:
            raise ValueError(f"Magnetizing inductance must be non-negative, got {inductance_mag}")
        if resistance_primary < 0:
            raise ValueError(f"Primary resistance must be non-negative, got {resistance_primary}")
        if resistance_secondary < 0:
            raise ValueError(f"Secondary resistance must be non-negative, got {resistance_secondary}")
        
        self.V1 = voltage_primary
        self.f = frequency
        self.n = turns_ratio  # N1/N2
        self.L_mag = inductance_mag
        self.R1 = resistance_primary
        self.R2 = resistance_secondary
        self.R_load = 100.0  # Default load on secondary
        
        self._update_omega()
    
    def _update_omega(self):
        """Update angular frequency."""
        self.omega = 2 * np.pi * self.f
    
    def set_primary_voltage(self, voltage):
        """Set primary voltage."""
        if voltage <= 0:
            raise ValueError(f"Primary voltage must be positive, got {voltage}")
        self.V1 = voltage
    
    def set_frequency(self, frequency):
        """Set frequency."""
        if frequency <= 0:
            raise ValueError(f"Frequency must be positive, got {frequency}")
        self.f = frequency
        self._update_omega()
    
    def set_turns_ratio(self, ratio):
        """Set turns ratio (N1/N2)."""
        if ratio <= 0:
            raise ValueError(f"Turns ratio must be positive, got {ratio}")
        self.n = ratio
    
    def set_magnetizing_inductance(self, inductance):
        """Set magnetizing inductance."""
        if inductance < 0:
            raise ValueError(f"Magnetizing inductance must be non-negative, got {inductance}")
        self.L_mag = inductance
    
    def set_primary_resistance(self, resistance):
        """Set primary winding resistance."""
        if resistance < 0:
            raise ValueError(f"Primary resistance must be non-negative, got {resistance}")
        self.R1 = resistance
    
    def set_secondary_resistance(self, resistance):
        """Set secondary winding resistance."""
        if resistance < 0:
            raise ValueError(f"Secondary resistance must be non-negative, got {resistance}")
        self.R2 = resistance
    
    def set_load(self, resistance):
        """Set load resistance on secondary."""
        if resistance <= 0:
            raise ValueError(f"Load resistance must be positive, got {resistance}")
        self.R_load = resistance
    
    def calculate_secondary_voltage_ideal(self):
        """Calculate ideal secondary voltage (no-load).
        
        Returns:
            float: Secondary voltage assuming ideal transformer
        """
        return self.V1 / self.n
    
    def calculate_magnetizing_current(self):
        """Calculate magnetizing current in primary.
        
        Returns:
            float: Magnetizing current in amperes
        """
        X_mag = self.omega * self.L_mag
        if X_mag == 0:
            return 0
        return self.V1 / X_mag
    
    def calculate_secondary_current(self):
        """Calculate secondary current.
        
        Returns:
            float: Secondary current in amperes
        """
        # Secondary voltage (ideal transformation minus secondary resistance drop)
        V2_ideal = self.V1 / self.n
        
        # Total secondary resistance
        R_sec_total = self.R2 + self.R_load
        
        # Secondary current (simplified - ignoring reflected impedance for now)
        I2 = V2_ideal / R_sec_total
        
        return I2
    
    def calculate_primary_current(self):
        """Calculate primary current.
        
        Returns:
            float: Primary current in amperes
        """
        # Secondary current
        I2 = self.calculate_secondary_current()
        
        # Reflected current from secondary (transformed)
        I_reflected = I2 / self.n
        
        # Magnetizing current
        I_mag = self.calculate_magnetizing_current()
        
        # Total primary current (vectorial sum - simplified as scalar for now)
        # In reality these add as phasors, but for educational purposes:
        I1 = np.sqrt(I_reflected**2 + I_mag**2)
        
        return I1
    
    def calculate_actual_secondary_voltage(self):
        """Calculate actual secondary voltage under load.
        
        Returns:
            float: Actual secondary voltage in volts
        """
        V2_ideal = self.V1 / self.n
        I2 = self.calculate_secondary_current()
        
        # Voltage drop in secondary winding
        V_drop_sec = I2 * self.R2
        
        # Actual voltage at load
        V2_actual = V2_ideal - V_drop_sec
        
        return V2_actual
    
    def get_all_values(self):
        """Calculate all transformer values.
        
        Returns:
            dict: Dictionary with all calculated values
        """
        # Voltages
        V1 = self.V1
        V2_ideal = self.calculate_secondary_voltage_ideal()
        V2_actual = self.calculate_actual_secondary_voltage()
        V_load = self.calculate_secondary_current() * self.R_load
        
        # Currents
        I1 = self.calculate_primary_current()
        I2 = self.calculate_secondary_current()
        I_mag = self.calculate_magnetizing_current()
        
        # Magnetizing reactance
        X_mag = self.omega * self.L_mag
        
        # Powers - Primary side
        S1 = V1 * I1  # Apparent power input
        
        # Powers - Secondary side (at load)
        P_load = I2**2 * self.R_load  # Active power to load
        P_loss_primary = I1**2 * self.R1  # Copper losses in primary
        P_loss_secondary = I2**2 * self.R2  # Copper losses in secondary
        
        # Reactive power (magnetizing)
        Q_mag = I_mag**2 * X_mag if X_mag > 0 else 0
        
        # Total apparent power on secondary
        S2 = V2_actual * I2
        
        # Power factor
        P_total = P_load + P_loss_primary + P_loss_secondary
        pf = P_total / S1 if S1 > 0 else 0
        
        # Efficiency
        efficiency = P_load / S1 if S1 > 0 else 0
        
        return {
            # Primary side
            'voltage_primary': V1,
            'current_primary': I1,
            'apparent_power_primary': S1,
            
            # Secondary side
            'voltage_secondary_ideal': V2_ideal,
            'voltage_secondary_actual': V2_actual,
            'voltage_load': V_load,
            'current_secondary': I2,
            'apparent_power_secondary': S2,
            
            # Magnetizing
            'current_magnetizing': I_mag,
            'reactance_magnetizing': X_mag,
            'reactive_power_magnetizing': Q_mag,
            
            # Power distribution
            'power_load': P_load,
            'power_loss_primary': P_loss_primary,
            'power_loss_secondary': P_loss_secondary,
            'power_total_input': P_total,
            
            # Performance
            'power_factor': pf,
            'efficiency': efficiency,
            'turns_ratio': self.n,
        }
    
    def get_parameters(self):
        """Get all transformer parameters.
        
        Returns:
            dict: Dictionary with all parameters
        """
        return {
            'voltage_primary': self.V1,
            'frequency': self.f,
            'turns_ratio': self.n,
            'inductance_magnetizing': self.L_mag,
            'resistance_primary': self.R1,
            'resistance_secondary': self.R2,
            'resistance_load': self.R_load,
        }
    
    def get_waveform_data(self, num_cycles=3, points_per_cycle=333):
        """Generate time-domain waveform data for voltage and current.
        
        Args:
            num_cycles: Number of AC cycles to generate (default: 3)
            points_per_cycle: Number of data points per cycle (default: 333, ~1000 total)
        
        Returns:
            dict: Dictionary containing:
                - 'time': Time array in seconds
                - 'v1': Primary voltage waveform (instantaneous values)
                - 'i1': Primary current waveform (instantaneous values)
                - 'v2': Secondary voltage waveform (instantaneous values)
                - 'i2': Secondary current waveform (instantaneous values)
        """
        # Get RMS values from existing calculations
        values = self.get_all_values()
        V1_rms = values['voltage_primary']
        I1_rms = values['current_primary']
        V2_rms = values['voltage_secondary_actual']
        I2_rms = values['current_secondary']
        pf = values['power_factor']
        
        # Calculate phase angle from power factor
        # For inductive load, current lags voltage
        # φ = arccos(pf), where pf = cos(φ)
        phi = np.arccos(np.clip(pf, -1.0, 1.0))  # Clip to avoid numerical errors
        
        # Generate time array
        period = 1.0 / self.f  # Period of one cycle
        t_end = num_cycles * period
        num_points = num_cycles * points_per_cycle
        time = np.linspace(0, t_end, num_points)
        
        # Angular frequency
        omega = 2 * np.pi * self.f
        
        # Generate waveforms (peak values = RMS × √2)
        # Primary voltage is the reference (0° phase)
        v1 = V1_rms * np.sqrt(2) * np.sin(omega * time)
        
        # Primary current lags voltage by phase angle φ
        i1 = I1_rms * np.sqrt(2) * np.sin(omega * time - phi)
        
        # Secondary voltage is in phase with primary (ideal transformer)
        # Uses actual secondary voltage (includes resistive drop)
        v2 = V2_rms * np.sqrt(2) * np.sin(omega * time)
        
        # Secondary current lags secondary voltage by same phase angle
        # (assuming load is resistive - same power factor on both sides)
        i2 = I2_rms * np.sqrt(2) * np.sin(omega * time - phi)
        
        return {
            'time': time,
            'v1': v1,
            'i1': i1,
            'v2': v2,
            'i2': i2,
        }
    
    def get_power_calculation_data(self, side='primary', num_cycles=3, points_per_cycle=333):
        """Generate power calculation data showing p(t) = v(t) × i(t).
        
        This method calculates instantaneous power and shows how integration
        produces real power P and reactive power Q.
        
        Args:
            side: Which side to calculate ('primary' or 'secondary')
            num_cycles: Number of AC cycles to generate (default: 3)
            points_per_cycle: Number of data points per cycle (default: 333)
        
        Returns:
            dict: Dictionary containing:
                - 'time': Time array in seconds
                - 'voltage': Voltage waveform v(t)
                - 'current': Current waveform i(t)
                - 'power_instantaneous': Instantaneous power p(t) = v(t) × i(t)
                - 'power_active': Average power P (real power)
                - 'power_reactive': Reactive power Q
                - 'power_apparent': Apparent power S
                - 'power_factor': Power factor cos(φ)
                - 'phase_angle': Phase angle φ in radians
        """
        # Get waveform data
        waveform_data = self.get_waveform_data(num_cycles, points_per_cycle)
        time = waveform_data['time']
        
        # Select voltage and current based on side
        if side.lower() == 'primary':
            voltage = waveform_data['v1']
            current = waveform_data['i1']
        elif side.lower() == 'secondary':
            voltage = waveform_data['v2']
            current = waveform_data['i2']
        else:
            raise ValueError(f"Invalid side '{side}'. Must be 'primary' or 'secondary'")
        
        # Calculate instantaneous power: p(t) = v(t) × i(t)
        power_instantaneous = voltage * current
        
        # Get RMS values and power metrics from transformer calculations
        values = self.get_all_values()
        pf = values['power_factor']
        
        if side.lower() == 'primary':
            V_rms = values['voltage_primary']
            I_rms = values['current_primary']
            S = values['apparent_power_primary']
        else:  # secondary
            V_rms = values['voltage_secondary_actual']
            I_rms = values['current_secondary']
            S = values['apparent_power_secondary']
        
        # Calculate powers
        # Active power (average of instantaneous power)
        P = np.mean(power_instantaneous)
        
        # From power triangle: P = S × cos(φ), Q = S × sin(φ)
        phi = np.arccos(np.clip(pf, -1.0, 1.0))
        Q = S * np.sin(phi)
        
        return {
            'time': time,
            'voltage': voltage,
            'current': current,
            'power_instantaneous': power_instantaneous,
            'power_active': P,
            'power_reactive': Q,
            'power_apparent': S,
            'power_factor': pf,
            'phase_angle': phi,
        }

File: web/src/models/Transformer.ts
import type {
  TransformerParams,
  TransformerValues,
  WaveformData,
  PowerCalculationData,
  TransformerSide,
  IMachine,
  MachineType,
  MachineMetadata,
} from '../types';
import {
  addHarmonicsToCurrentArray,
  calculateHarmonicMetrics,
} from '../utils/harmonics';

/**
 * AC Transformer model with full electrical calculations
 * 
 * This class implements the physics of a single-phase AC transformer
 * including voltage transformation, current calculations, power analysis,
 * and time-domain waveform generation.
 * 
 * Implements IMachine interface for compatibility with multi-machine platform.
 */
export class Transformer implements IMachine {
  readonly type: MachineType = 'transformer';
  private params: TransformerParams;

  constructor(params: TransformerParams) {
    // Ensure voltage alias matches voltagePrimary
    const voltage = params.voltagePrimary || params.voltage;
    this.params = { 
      ...params,
      voltage,
      voltagePrimary: voltage,
    };
  }

  /**
   * Get machine metadata for UI rendering
   */
  getMetadata(): MachineMetadata {
    return {
      name: 'AC Transformer',
      description: 'Single-phase power transformer with magnetizing inductance',
      icon: '⚡',
      category: 'static',
      supportsWaveforms: true,
      supportsPowerTriangle: true,
    };
  }

  /**
   * Update transformer parameters
   */
  updateParams(params: Partial<TransformerParams>): void {
    this.params = { ...this.params, ...params } as TransformerParams;
    // Keep voltage alias in sync with voltagePrimary
    if (params.voltagePrimary !== undefined) {
      this.params.voltage = params.voltagePrimary;
    } else if (params.voltage !== undefined) {
      this.params.voltagePrimary = params.voltage;
    }
  }

  /**
   * Get current parameters
   */
  getParams(): TransformerParams {
    return { ...this.params };
  }

  /**
   * Calculate all transformer values
   */
  calculate(): TransformerValues {
    const {
      voltagePrimary: V1,
      frequency: f,
      turnsRatio: n,
      inductanceMag: Lmag,
      resistancePrimary: R1,
      resistanceSecondary: R2,
      resistanceLoad: Rload,
    } = this.params;

    // Angular frequency
    const omega = 2 * Math.PI * f;

    // Magnetizing reactance
    const Xmag = omega * Lmag;

    // Iterative solution needed since I1 affects Vmag which affects Imag which affects I1
    // Start with approximation
    let I1 = 0.1; // Initial guess
    let I2_reflected = 0;
    let Imag = 0;
    let Vmag = V1;
    let V2 = V1 / n;
    let I2 = 0;
    
    // Iterate to converge (usually 3-5 iterations enough)
    for (let iter = 0; iter < 10; iter++) {
      // Magnetizing voltage after R1 drop
      Vmag = V1 - I1 * R1;
      
      // Secondary voltage (ideal transformation from magnetizing voltage)
      V2 = Vmag / n;
      
      // Load current
      I2 = V2 / (R2 + Rload);
      
      // Current reflected to primary
      I2_reflected = I2 / n;
      
      // Magnetizing current
      Imag = Vmag / Xmag;
      
      // Total primary current (phasor sum)
      const I1_new = Math.sqrt(I2_reflected ** 2 + Imag ** 2);
      
      // Check convergence
      if (Math.abs(I1_new - I1) < 0.0001) {
        I1 = I1_new;
        break;
      }
      I1 = I1_new;
    }
    
    const Iload = I2;

    // Power calculations - Primary side
    const P1 = V1 * I2_reflected;  // Active power (resistive component)
    const Q1 = V1 * Imag;          // Reactive power (magnetizing)
    const S1 = V1 * I1;            // Apparent power

    // Power calculations - Secondary side
    const P2 = V2 * I2;
    const Q2 = 0;  // No reactive component on secondary (resistive load)
    const S2 = V2 * I2;

    // Load power
    const Pload = I2 ** 2 * Rload;

    // Magnetizing reactive power
    const Qmag = Q1;

    // Power factor and phase angle (displacement, from fundamental only)
    const powerFactor = P1 / S1;
    // Phase angle for inductive load (negative because current lags voltage)
    const phaseAngle = -Math.acos(powerFactor);
    
    // Efficiency (output power / input power)
    const efficiency = P2 / P1;

    // Harmonic metrics
    const h3 = (this.params.harmonic3 as number) ?? 0;
    const h5 = (this.params.harmonic5 as number) ?? 0;
    const h7 = (this.params.harmonic7 as number) ?? 0;
    const harmonics = calculateHarmonicMetrics(I1, V1, P1, Q1, h3, h5, h7);
    const powerApparentTrue = harmonics.S_total;
    const powerFactorTrue = harmonics.truePF;

    return {
      voltageSecondary: V2,
      voltageMagnetizing: Vmag,
      currentPrimary: harmonics.I_total_rms,
      currentSecondary: I2,
      currentMagnetizing: Imag,
      currentLoad: Iload,
      powerActivePrimary: P1,
      powerReactivePrimary: Q1,
      powerApparentPrimary: powerApparentTrue,
      powerActiveSecondary: P2,
      powerReactiveSecondary: Q2,
      powerApparentSecondary: S2,
      powerLoad: Pload,
      powerReactiveMagnetizing: Qmag,
      powerFactor: powerFactorTrue,
      phaseAngle,
      // Base interface values (for IMachine compatibility)
      powerActive: P1,
      powerReactive: Q1,
      powerApparent: powerApparentTrue,
      efficiency,
      // Harmonic distortion metrics
      thdCurrent: harmonics.thd,
      powerDistortion: harmonics.powerDistortion,
      displacementPowerFactor: harmonics.displacementPF,
      distortionPowerFactor: harmonics.distortionPF,
      truePowerFactor: harmonics.truePF,
    };
  }

  /**
   * Generate time-domain waveform data
   * 
   * Creates voltage and current waveforms for both primary and secondary sides
   * over 3-4 AC cycles for visualization.
   * 
   * @param cycles - Number of AC cycles to generate (default: 3)
   * @param pointsPerCycle - Number of data points per cycle (default: 100)
   */
  getWaveformData(cycles: number = 3, pointsPerCycle: number = 100): WaveformData {
    const { voltagePrimary: V1, frequency: f } = this.params;
    const values = this.calculate();
    
    const V2 = values.voltageSecondary;
    const phi = values.phaseAngle;

    // Time array
    const totalPoints = cycles * pointsPerCycle;
    const period = 1 / f;
    const dt = (cycles * period) / totalPoints;
    const time: number[] = [];
    
    for (let i = 0; i < totalPoints; i++) {
      time.push(i * dt);
    }

    // Angular frequency
    const omega = 2 * Math.PI * f;

    // Harmonic amplitudes — needed to recover the fundamental current from I_total_rms
    const h3 = (this.params.harmonic3 as number) ?? 0;
    const h5 = (this.params.harmonic5 as number) ?? 0;
    const h7 = (this.params.harmonic7 as number) ?? 0;
    const thd = Math.sqrt(h3 ** 2 + h5 ** 2 + h7 ** 2);

    // values.currentPrimary is I_total_rms (what a clamp meter would read).
    // Recover the fundamental RMS so the base sinusoid carries correct P.
    // I_fund = I_total / √(1 + THD²)
    const I1_fund = (values.currentPrimary as number) / Math.sqrt(1 + thd ** 2);
    const I2_fund = (values.currentSecondary as number); // secondary not inflated

    // Peak values (RMS × √2)
    const V1_peak = V1 * Math.sqrt(2);
    const V2_peak = V2 * Math.sqrt(2);
    const I1_peak = I1_fund * Math.sqrt(2);   // fundamental peak
    const I2_peak = I2_fund * Math.sqrt(2);

    // Generate waveforms
    const v1 = time.map(t => V1_peak * Math.sin(omega * t));
    const v2 = time.map(t => V2_peak * Math.sin(omega * t));

    // Fundamental current waveforms
    const i1_fund = time.map(t => I1_peak * Math.sin(omega * t - phi));
    const i2_fund = time.map(t => I2_peak * Math.sin(omega * t - phi));

    // Apply current harmonics (harmonic amplitudes scale from the fundamental peak)
    const i1 = addHarmonicsToCurrentArray(time, i1_fund, omega, h3, h5, h7, I1_peak);
    const i2 = addHarmonicsToCurrentArray(time, i2_fund, omega, h3, h5, h7, I2_peak);

    return { time, v1, i1, v2, i2 };
  }

  /**
   * Generate power calculation data showing p(t) = v(t) × i(t)
   * 
   * Calculates instantaneous power and demonstrates how active (P),
   * reactive (Q), and apparent (S) power are derived from the waveforms.
   * 
   * @param side - Which side to calculate ('primary' or 'secondary')
   * @param cycles - Number of AC cycles (default: 3)
   * @param pointsPerCycle - Points per cycle (default: 100)
   */
  /**
   * Get power calculation data (IMachine interface implementation)
   * @param points - Number of points to generate (uses pointsPerCycle internally)
   */
  getPowerCalculationData(points?: number): PowerCalculationData {
    const pointsPerCycle = points ? Math.floor(points / 3) : 100;
    return this.getPowerCalculationDataForSide('primary', 3, pointsPerCycle);
  }

  /**
   * Get power calculation data for specific transformer side
   * @param side - Which side to calculate power for
   * @param cycles - Number of AC cycles
   * @param pointsPerCycle - Points per cycle
   */
  getPowerCalculationDataForSide(
    side: TransformerSide = 'primary',
    cycles: number = 3,
    pointsPerCycle: number = 100
  ): PowerCalculationData {
    const waveforms = this.getWaveformData(cycles, pointsPerCycle);
    const values = this.calculate();
    const { frequency: f } = this.params;

    // Select voltage and current based on side
    const voltage = side === 'primary' ? waveforms.v1 : waveforms.v2;
    const current = side === 'primary' ? waveforms.i1 : waveforms.i2;

    // Calculate instantaneous power p(t) = v(t) × i(t)
    const powerInstantaneous = voltage.map((v, i) => v * current[i]);

    // Active power = average of instantaneous power
    const powerActive = powerInstantaneous.reduce((sum, p) => sum + p, 0) / powerInstantaneous.length;
    
    // Apparent power and power factor
    const powerApparent = side === 'primary'
      ? values.powerApparentPrimary
      : values.powerApparentSecondary;

    const powerFactor = values.powerFactor;

    // Use displacement reactive power directly — it equals V×I×sin(φ₁) and is
    // what the Power Triangle also shows. Computing Q as S×sin(acos(truePF))
    // would give √(Q²+D²) which conflates reactive and distortion power.
    const powerReactive = side === 'primary'
      ? values.powerReactivePrimary
      : values.powerReactiveSecondary;

    // Generate pure magnetizing power waveform (only for primary side)
    // p_mag(t) = v(t) × i_mag(t) where i_mag lags v by 90°
    const powerMagnetizing = side === 'primary' ? (() => {
      const omega = 2 * Math.PI * f;
      const V1_peak = this.params.voltagePrimary * Math.sqrt(2);
      const Imag_peak = values.currentMagnetizing * Math.sqrt(2);
      return waveforms.time.map(t => 
        V1_peak * Math.sin(omega * t) * Imag_peak * Math.sin(omega * t - Math.PI / 2)
      );
    })() : null;

    return {
      time: waveforms.time,
      voltage,
      current,
      powerInstantaneous,
      powerActive,
      powerReactive,
      powerApparent,
      powerMagnetizing,
      powerFactor,
    };
  }
}
